url20b0f880a5_jgraph_legacy_jgraph5_tgz-pJ8-org_jgraph_example_HelloWorldJ8/src/com/mxgraph/canvas/mxSvgCanvas.java

```diff
@@ -0,0 +1,774 @@
+	 */
+	public void setDocument(Document document)
+	{
+		this.document = document;
+	}
+
+	/**
+	 * Returns a reference to the document that represents the canvas.
+	 * 
+	 * @return Returns the document.
+	 */
+	public Document getDocument()
+	{
+		return document;
+	}
+
+	/*
+	 * (non-Javadoc)
+	 * @see com.mxgraph.canvas.mxICanvas#drawCell()
+	 */
+	public Object drawCell(mxCellState state)
+	{
+		Map<String, Object> style = state.getStyle();
+		Element elem = null;
+
+		if (state.getAbsolutePointCount() > 1)
+		{
+			List<mxPoint> pts = state.getAbsolutePoints();
+
+			// Transpose all points by cloning into a new array
+			pts = mxUtils.translatePoints(pts, translate.x, translate.y);
+
+			// Draws the line
+			elem = drawLine(pts, style);
+
+			// Applies opacity
+			float opacity = mxUtils.getFloat(style, mxConstants.STYLE_OPACITY,
+					100);
+
+			if (opacity != 100)
+			{
+				String value = String.valueOf(opacity / 100);
+				elem.setAttribute("fill-opacity", value);
+				elem.setAttribute("stroke-opacity", value);
+			}
+		}
+		else
+		{
+			int x = (int) state.getX() + translate.x;
+			int y = (int) state.getY() + translate.y;
+			int w = (int) state.getWidth();
+			int h = (int) state.getHeight();
+
+			if (!mxUtils.getString(style, mxConstants.STYLE_SHAPE, "").equals(
+					mxConstants.SHAPE_SWIMLANE))
+			{
+				elem = drawShape(x, y, w, h, style);
+			}
+			else
+			{
+				int start = (int) Math.round(mxUtils.getInt(style,
+						mxConstants.STYLE_STARTSIZE,
+						mxConstants.DEFAULT_STARTSIZE)
+						* scale);
+
+				// Removes some styles to draw the content area
+				Map<String, Object> cloned = new Hashtable<String, Object>(
+						style);
+				cloned.remove(mxConstants.STYLE_FILLCOLOR);
+				cloned.remove(mxConstants.STYLE_ROUNDED);
+
+				if (mxUtils.isTrue(style, mxConstants.STYLE_HORIZONTAL, true))
+				{
+					elem = drawShape(x, y, w, start, style);
+					drawShape(x, y + start, w, h - start, cloned);
+				}
+				else
+				{
+					elem = drawShape(x, y, start, h, style);
+					drawShape(x + start, y, w - start, h, cloned);
+				}
+			}
+		}
+
+		return elem;
+	}
+
+	/*
+	 * (non-Javadoc)
+	 * @see com.mxgraph.canvas.mxICanvas#drawLabel()
+	 */
+	public Object drawLabel(String label, mxCellState state, boolean html)
+	{
+		mxRectangle bounds = state.getLabelBounds();
+
+		if (drawLabels)
+		{
+			int x = (int) bounds.getX() + translate.x;
+			int y = (int) bounds.getY() + translate.y;
+			int w = (int) bounds.getWidth();
+			int h = (int) bounds.getHeight();
+			Map<String, Object> style = state.getStyle();
+
+			return drawText(label, x, y, w, h, style);
+		}
+
+		return null;
+	}
+
+	/**
+	 * Draws the shape specified with the STYLE_SHAPE key in the given style.
+	 * 
+	 * @param x X-coordinate of the shape.
+	 * @param y Y-coordinate of the shape.
+	 * @param w Width of the shape.
+	 * @param h Height of the shape.
+	 * @param style Style of the the shape.
+	 */
+	public Element drawShape(int x, int y, int w, int h,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 183 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 99% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 421-492 into a separate method.